### ISSUE TITLE
rtl8812au Compile Error "rtw_select_queue"

### ISSUE THREAD

**Issue by jtome7** [aircrack-ng/rtl8812au#861]
May I please ask if there's an easy way to get around this compile error?
Using CentOS  ( which is like Fedora & Red Hat ) kernel 4.18

/var/lib/dkms/rtl8812au/5.2.20.2/build/os_dep/linux/os_intfs.c:1393:22: error: initialization of ‘u16 (*)(struct net_device *, struct sk_buff *, struct net_device *, u16 (*)(struct net_device *, struct sk_buff *, struct net_device *))’ {aka ‘short unsigned int (*)(struct net_device *, struct sk_buff *, struct net_device *, short unsigned int (*)(struct net_device *, struct sk_buff *, struct net_device *))’} from incompatible pointer type ‘u16 (*)(struct net_device *, struct sk_buff *, void *, u16 (*)(struct net_device *, struct sk_buff *, struct net_device *))’ {aka ‘short unsigned int (*)(struct net_device *, struct sk_buff *, void *, short unsigned int (*)(struct net_device *, struct sk_buff *, struct net_device *))’} [-Werror=incompatible-pointer-types]
  .ndo_select_queue = rtw_select_queue,
                      ^~~~~~~~~~~~~~~~
/var/lib/dkms/rtl8812au/5.2.20.2/build/os_dep/linux/os_intfs.c:1393:22: note: (near initialization for ‘rtw_netdev_ops.ndo_select_queue’)

**Comment by CGarces** [comment]
CentOS should work after #599 on branch v5.6.4.2
¿Are you using a different branch?

Also check issue #542 for some [alternatives ](https://github.com/aircrack-ng/rtl8812au/issues/542#issuecomment-587775745)like using ELRepo's kernels

Duplicate of #542

**Comment by jtome7** [comment]
Hello, Carlos!

There are millions of Linux users world-wide on the distribution family of
Red Hat / CentOS / Fedora Linux, so thanks for adding support for Realtek
8812 & 8814 chipset for those millions of users!

The title says rtl8812au ... does it also work for rtl8814au that is found
in quite a number of wifi adapter products like this one:

TP-Link AC1900 Archer T9UH

[image: TP-Link-Archer-T9UH-WIFI-Adapter.jpg]



*Thanks!**Joel*


*************************** Joel Tome *

*jtome7 @ gmail.com <http://gmail.com>****************************

**Comment by CGarces** [comment]
Support for the rtl8814au chipset was removed from this driver. Use https://github.com/aircrack-ng/rtl8814au/

**Comment by jtome7** [comment]
Thanks, I'll check it out.



*Thanks!**Joel*


*************************** Joel Tome *

*jtome7 @ gmail.com <http://gmail.com>****************************

**Comment by jtome7** [comment]
Just to let folks know ... I tried several 5.x kernels from the ElRepo ... nothing worked on my CentOS 8.3 with my TP-Link AC1900 Archer t9UH wifi adapter with Realtek 8814 chipset.  Thanks for continuing suggestions as new updates become available.  Appreciate your efforts!

**Comment by lepp9922** [comment]
I can also confirm the same error, compiling fails on the latest AlmaLinux 8 (RHEL/CentOS 8). Kernel 4.18

**Comment by jtome7** [comment]
The crazy thing is that just before the 8814au driver was separated from this rtl8812au branch,  the TP-Link AC1900 Archer t9UH wifi adapter with Realtek 8814 chipset WORKED!  I wish there were a way to get that previous compilation back that contained both 8812 & 8814.

**Comment by CGarces** [comment]
I have fixed the compilation
https://github.com/aircrack-ng/rtl8812au/compare/v5.6.4.2...CGarces:centos
Tested on CentOS 8 VM, not on real hardware.

**Comment by Rondore** [comment]
The patch by @CGarces allowed me to compile this driver for Rocky Linux 8.4 running kernel 4.18.0-305.3.1.el8_4.x86_64. I ran this test on real hardware and my WiFi is now working great. Thanks!

**Comment by kimocoder** [comment]
rtl8814au support is added back and fixes has been pushed. Update and build/install.

If problem persists, open this issue report again :+1:

**Comment by fs30000** [comment]
> I have fixed the compilation [v5.6.4.2...CGarces:centos](https://github.com/aircrack-ng/rtl8812au/compare/v5.6.4.2...CGarces:centos) Tested on CentOS 8 VM, not on real hardware.

Repo cloned today 13-01-2023. Rocky 8.7 kernel  4.18.0-425.3.1.el8.x86_64

Lines applied on lines 7775 and 10229, but then it fails with:

```
/var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.c: In function ‘rtw_cfg80211_ch_switch_notify’:
/var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.c:468:3: error: too few arguments to function ‘cfg80211_ch_switch_started_notify’
   cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/include/osdep_service_linux.h:97,
                 from /var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/include/osdep_service.h:58,
                 from /var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/include/drv_types.h:26,
                 from /var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.c:17:
./include/net/cfg80211.h:7902:6: note: declared here
 void cfg80211_ch_switch_started_notify(struct net_device *dev,
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.c: At top level:
/var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.c:10232:3: error: ‘struct cfg80211_ops’ has no member named ‘mgmt_frame_register’
  .mgmt_frame_register = cfg80211_rtw_mgmt_frame_register,
   ^~~~~~~~~~~~~~~~~~~
/var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.c:10232:25: error: initialization of ‘int (*)(struct wiphy *, struct wireless_dev *, u64)’ {aka ‘int (*)(struct wiphy *, struct wireless_dev *, long long unsigned int)’} from incompatible pointer type ‘void (*)(struct wiphy *, struct wireless_dev *, u16,  bool)’ {aka ‘void (*)(struct wiphy *, struct wireless_dev *, short unsigned int,  _Bool)’} [-Werror=incompatible-pointer-types]
  .mgmt_frame_register = cfg80211_rtw_mgmt_frame_register,
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.c:10232:25: note: (near initialization for ‘rtw_cfg80211_ops.mgmt_tx_cancel_wait’)
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:317: /var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build/os_dep/linux/ioctl_cfg80211.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:1585: _module_/var/lib/dkms/rtl88x2bu/5.6.4.2~20221105/build] Error 2
make[1]: Leaving directory '/usr/src/kernels/4.18.0-425.3.1.el8.x86_64'
make: *** [Makefile:1730: modules] Error 2

```

**Comment by fs30000** [comment]
> I have fixed the compilation [v5.6.4.2...CGarces:centos](https://github.com/aircrack-ng/rtl8812au/compare/v5.6.4.2...CGarces:centos) Tested on CentOS 8 VM, not on real hardware.

CGarces, i have Rocky 8.7, it fails to compile as well. I don't know much about git, it there a way to clone your repo with all the changes for Centos/Rocly/Redhat?
